### PR TITLE
upper_bound and lower_bound functions added

### PR DESCRIPTION
**Pull request by sHiVaNgI821**
#### References to other Issues or PRs or Relevant literature

Added the upper_bound and lower_bound functions to the algorithms library. Upper_bound returns the index of the first occurence of an element greater than value, in the given sorted OneDimensionalArray. Lower_bound returns the index of the first occurence of an element which is not less than value, in the given OneDimensionalArray.

### CODE REVIEW

**Comment by Smit-create** [comment]
@sHiVaNgI821 Please address this https://github.com/codezonediitj/pydatastructs/pull/344#issuecomment-797416009 in your PR too

File: pydatastructs/linear_data_structures/algorithms.py [hunk 1]
```diff
@@ -787,3 +789,108 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+    >>> upperBound
+    4
+
+    Note
+    ====
+
+    The OneDimensionalArray must be sorted beforehand
+    """
+    if comp is None:
+        comp = lambda a, b: (a <= b)
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:832]: The order needs to be well defined. i.e, strict inequality is followed in order. You can have a look at the implementation here: https://en.cppreference.com/w/cpp/algorithm/upper_bound.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 2]
```diff
@@ -787,3 +789,108 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+    index = end
+    if not comp(array[start], value):
+        index = start
+    while start <= end:
+        mid = (start+end)//2
+        if comp(array[mid],value):
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:838]:
```suggestion
        mid = (start + end)//2
        if comp(array[mid], value):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 3]
```diff
@@ -787,3 +789,108 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+    Parameters
+    ========
+    array: OneDimensionalArray
+    The sorted array (sorted according to a custom comparator function) in which the lower bound has to be found
+
+    start: int
+    The staring index of the portion of the array in which the lower bound of a given value has to be looked for
+
+    end: int
+    The ending index of the portion of the array in which the lower bound of a given value has to be looked for
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:858]: The line width should not exceed 79 characters. Wrap up text in multiple lines not more than 79 chars

File: pydatastructs/linear_data_structures/algorithms.py [hunk 4]
```diff
@@ -787,3 +789,108 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+
+    Returns
+    =======
+
+    output: int
+    Lower bound of the given value in the given sorted OneDimensionalArray
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:864]:
```suggestion
    output: int
        Lower bound of the given value in the given sorted OneDimensionalArray
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 5]
```diff
@@ -787,3 +789,108 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+    The OneDimensionalArray must be sorted beforehand
+    """
+    if comp is None:
+        comp = lambda a, b: (a < b)
+    index = end
+    if not comp(array[start], value):
+        index = start
+    while start <= end:
+        mid = (start+end)//2
+        if comp(array[mid],value):
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:891]:
```suggestion
        mid = (start + end)//2
        if comp(array[mid], value):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 6]
```diff
@@ -787,3 +789,108 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+            start = mid + 1
+        else:
+            index = mid
+            end = mid -1
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:895]:
```suggestion
            end = mid - 1
```

**Review by Smit-create — COMMENTED** [review]
This looks good. Just have a look at some suggestions.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 7]
```diff
@@ -787,3 +789,116 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+
+    end: int
+        The ending index of the portion of the array in which the upper bound
+        of a given value has to be looked for
+
+    Returns
+    =======
+
+    output: int
+        Upper bound of the given value in the given sorted OneDimensionalArray
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:816]: Index of upper bound of given value...

File: pydatastructs/linear_data_structures/algorithms.py [hunk 8]
```diff
@@ -787,3 +789,116 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+
+    end: int
+        The ending index of the portion of the array in which the lower
+        bound of a given value has to be looked for
+
+    Returns
+    =======
+
+    output: int
+        Lower bound of the given value in the given sorted OneDimensionalArray
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:872]: Same

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 9]
```diff
@@ -126,3 +127,78 @@ def test_longest_common_sequence():
     Z = ODA(int, [])
     output = longest_common_subsequence(Y, Z)
     assert str(output) == '[]'
+
+def test_upper_bound():
+    ODA = OneDimensionalArray
+    arr1 = ODA(int, [3, 3, 3])
+    output = upper_bound(arr1, 0, len(arr1) - 1, 3, None)
+    expected_result = 2
+    assert expected_result == output
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:136]: From, here I think there needs a slight change in the algo. That is output `2` isn't correct mathematically for this in the given range. Hence, we should consider the interval as [start, end). And so, I expect the output of it as `3`. (You can verify it by yourself, we are making it similar to cpp)

Comment (Smit-create) [reply pydatastructs/linear_data_structures/tests/test_algorithms.py:136]: Also, after making this interval change, please add tests apart from [0, len(arr)) interval too.

**Comment by Smit-create** [comment]
Looks good to me.

**Comment by Smit-create** [comment]
@czgdp1807 Have a look at it

**Comment by Smit-create** [comment]
@sHiVaNgI821 Please resolve merge conflicts

File: pydatastructs/linear_data_structures/algorithms.py [hunk 10]
```diff
@@ -843,3 +845,123 @@ def is_ordered(array, **kwargs):
...
+        The sorted array (sorted according to a custom comparator function) in which the
+        upper bound has to be found
+
+    start: int
+        The staring index of the portion of the array in which the upper bound
+        of a given value has to be looked for
+
+    end: int
+        The ending index of the portion of the array in which the upper bound
+        of a given value has to be looked for
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:866]:
```suggestion
    start: int, optional.
        The staring index of the portion of the array in which the upper bound
        of a given value has to be looked for. By default is 0.

    end: int, optional
        The ending index of the portion of the array in which the upper bound
        of a given value has to be looked for. By default is len(array)
    
    comp: boolean funtion, optional
        A function that takes two paramters and returns True if 1st parameter is strictly
        smaller than 2nd parameter
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 11]
```diff
@@ -843,3 +845,123 @@ def is_ordered(array, **kwargs):
...
+    Note
+    ====
+
+    The OneDimensionalArray must be sorted beforehand
+    """
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array))
+    comp = kwargs.get('comp', lambda x,y: x < y)
+    # if comp is None:
+    #     comp = lambda a, b: (a < b)
```
Comment (Smit-create) [inline pydatastructs/linear_data_structures/algorithms.py:895]: Remove this comment

**Review by Smit-create — COMMENTED** [review]
Apply these suggestions to both functions. Apart from this, it looks good to me.

**Comment by czgdp1807** [comment]
Looks Good. Will polish it and push back again.

**Review by Smit-create — APPROVED** [review]

**Comment by Smit-create** [comment]
@czgdp1807 This can be merged.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 12]
```diff
@@ -843,3 +845,144 @@ def is_ordered(array, **kwargs):
...
+    1
+    >>> arr = ODA(int, [7, 6, 5, 5, 4])
+    >>> lb = lower_bound(arr, 5, start=0, comp=lambda x, y : x > y)
+    >>> lb
+    2
+
+    Note
+    ====
+
+    The given OneDimensionalArray object must be sorted.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:971]:
```suggestion
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 13]
```diff
@@ -843,3 +845,144 @@ def is_ordered(array, **kwargs):
...
+        else:
+            index = mid
+            inclusive_end = mid - 1
+    return index
+
+def lower_bound(array, value, **kwargs):
+    """
+    Finds the the index of the first occurence of an element which is not
+    less than the given value according to specified order,
+    in the given sorted OneDimensionalArray
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:923]:
```suggestion
    in the given OneDimensionalArray.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 14]
```diff
@@ -843,3 +845,144 @@ def is_ordered(array, **kwargs):
...
+
+    Parameters
+    ==========
+
+    array: OneDimensionalArray
+        The sorted array in which the lower bound has to be found.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:929]:
```suggestion
        The array in which the lower bound has to be found.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 15]
```diff
@@ -843,3 +845,144 @@ def is_ordered(array, **kwargs):
         if comp(array[i], array[i - 1]):
             return False
     return True
+
+def upper_bound(array, value, **kwargs):
+    """
+    Finds the index of the first occurence of an element greater than the given
+    value according to specified order, in the given sorted OneDimensionalArray.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:852]:
```suggestion
    value according to specified order, in the given OneDimensionalArray.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 16]
```diff
@@ -843,3 +845,144 @@ def is_ordered(array, **kwargs):
...
+
+    Parameters
+    ==========
+
+    array: OneDimensionalArray
+        The sorted array in which the upper bound has to be found.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:858]:
```suggestion
        The array in which the upper bound has to be found.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 17]
```diff
@@ -843,3 +845,144 @@ def is_ordered(array, **kwargs):
...
+        for specifying the desired ordering.
+        Optional, by default, less than or
+        equal to is used for comparing two
+        values.
+
+    Returns
+    =======
+
+    output: int
+        Index of the upper bound of the given value in the given sorted OneDimensionalArray
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:882]:
```suggestion
        Index of the upper bound of the given value in the given OneDimensionalArray.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 18]
```diff
@@ -843,3 +845,144 @@ def is_ordered(array, **kwargs):
...
+    3
+    >>> arr2 = ODA(int, [7, 6, 5, 5, 4])
+    >>> ub = upper_bound(arr2, 5, comp=lambda x, y: x > y)
+    >>> ub
+    4
+
+    Note
+    ====
+
+    The given OneDimensionalArray object must be sorted.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:900]:
```suggestion
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 19]
```diff
@@ -843,3 +845,142 @@ def is_ordered(array, **kwargs):
...
+
+    end: int, optional
+        The ending index of the portion of the array in which the upper bound
+        of a given value has to be looked for.
+        Optional, by default the index
+        of the last position filled.
+
+    comp: lambda/function
+        The comparator which is to be used
+        for specifying the desired ordering.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:873]:
```suggestion
    array: OneDimensionalArray
        The array in which the upper bound has to be found.
    start: int
        The staring index of the portion of the array in which the upper bound
        of a given value has to be looked for.
        Optional, by default 0
    end: int, optional
        The ending index of the portion of the array in which the upper bound
        of a given value has to be looked for.
        Optional, by default the index
        of the last position filled.
    comp: lambda/function
        The comparator which is to be used
        for specifying the desired ordering.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 20]
```diff
@@ -843,3 +845,142 @@ def is_ordered(array, **kwargs):
...
+        Optional, by default 0
+
+    end: int, optional
+        The ending index of the portion of the array in which the upper bound
+        of a given value has to be looked for.
+        Optional, by default the index
+        of the last position filled.
+
+    comp: lambda/function
+        The comparator which is to be used
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:942]:
```suggestion

    array: OneDimensionalArray
        The array in which the lower bound has to be found.
    start: int
        The staring index of the portion of the array in which the upper bound
        of a given value has to be looked for.
        Optional, by default 0
    end: int, optional
        The ending index of the portion of the array in which the upper bound
        of a given value has to be looked for.
        Optional, by default the index
        of the last position filled.
    comp: lambda/function
        The comparator which is to be used
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 21]
```diff
@@ -843,3 +845,142 @@ def is_ordered(array, **kwargs):
...
+        for specifying the desired ordering.
+        Optional, by default, less than or
+        equal to is used for comparing two
+        values.
+
+    Returns
+    =======
+
+    output: int
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:951]:
```suggestion
    index: int
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 22]
```diff
@@ -843,3 +845,142 @@ def is_ordered(array, **kwargs):
...
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:881]:
```suggestion
    index: int
```

**Comment by czgdp1807** [comment]
Thanks.

**Pull request merged**